File: src/main/java/org/gga/graph/transform/Morph.java
package org.gga.graph.transform;

import com.google.common.base.Function;
import com.google.common.collect.Iterables;
import com.google.common.collect.ObjectArrays;
import org.gga.graph.Edge;
import org.gga.graph.Graph;
import org.gga.graph.MutableGraph;
import org.gga.graph.impl.DataGraphImpl;
import org.gga.graph.impl.SparseGraphImpl;
import org.gga.graph.maps.DataGraph;
import org.gga.graph.util.IntIntFunction;

import java.util.List;
import java.util.Map;

import static com.google.common.base.Preconditions.checkState;
import static com.google.common.collect.Lists.newArrayList;
import static com.google.common.collect.Maps.newHashMap;
import static com.google.common.collect.Maps.newLinkedHashMap;

/**
 * @author mike
 */
public final class Morph {
    public static Graph morphGraph(Graph g, IntIntFunction map) {
        int[] vertexMap = new int[g.V()];
        int maxVertex = -1;

        for (int v = 0; v < g.V(); v++) {
            int t = map.apply(v);
            vertexMap[v] = t;
            maxVertex = Math.max(maxVertex, t);
        }

        int newSize = maxVertex + 1;
        checkState(newSize > 0);

        MutableGraph result = new SparseGraphImpl(newSize, g.isDirected());

        for (int v = 0; v < g.V(); v++) {
            for (Edge e : g.getEdges(v)) {
                int w = e.w();

                int v1 = vertexMap[v];
                int w1 = vertexMap[w];

                if (result.edge(v1, w1) == null) {
                    result.insert(v1, w1);
                }
            }
        }

        return result;
    }

    public static <N, E> DataGraph<N, E> isomorph(DataGraph<N, E> g, IntIntFunction map) {
        int[] vertexMap = new int[g.V()];
        int newSize = -1;

        for (int v = 0; v < g.V(); v++) {
            int t = map.apply(v);
            vertexMap[v] = t;
            newSize = Math.max(newSize, t);
        }

        newSize++;
        checkState(newSize > 0 && newSize == g.V());

        DataGraph<N, E> result = new DataGraphImpl<N, E>(g.getNodeClass(), newSize, g.isDirected());

        for (int v = 0; v < g.V(); v++) {
            result.setNode(vertexMap[v], g.getNode(v));
        }

        for (int v = 0; v < g.V(); v++) {
            for (Edge e : g.getIntGraph().getEdges(v)) {
                int w = e.w();

                int v1 = vertexMap[v];
                int w1 = vertexMap[w];

                result.insert(v1, w1, g.edge(v, w));
            }
        }

        return result;
    }

    public static <N, E, N1, E1> DataGraph<N1, E1> morph(
            DataGraph<N, E> g,
            Function<N, N1> nodeMap,
            Function<List<E>, E1> verticesMap,
            Class<N1> newNodeClass,
            boolean allowSelfLoops) {
        N1[] nodesDataMap = ObjectArrays.newArray(newNodeClass, g.V());

        for (int v = 0; v < g.V(); v++) {
            nodesDataMap[v] = nodeMap.apply(g.getNode(v));
        }

        Map<N1, Map<N1, List<E>>> newEdges = newLinkedHashMap();

        for (int v = 0; v < g.V(); v++) {
            N1 n1 = nodesDataMap[v];

            Map<N1, List<E>> edges = newEdges.get(n1);
            if (edges == null) {
                edges = newHashMap();
                newEdges.put(n1, edges);
            }

            for (Edge e : g.getIntGraph().getEdges(v)) {
                int w = e.w();

                N1 n2 = nodesDataMap[w];
                List<E> list = edges.get(n2);
                if (list == null) {
                    list = newArrayList();
                    edges.put(n2, list);
                }

                list.add(g.getEdge(e));
            }
        }

        N1[] newNodes = Iterables.toArray(newEdges.keySet(), newNodeClass);
        int newSize = newNodes.length;

        DataGraph<N1, E1> result = new DataGraphImpl<N1,E1>(newNodeClass, newSize, g.isDirected());

        for (int v = 0; v < newSize; v++) {
            result.setNode(v, newNodes[v]);
        }

        for (Map.Entry<N1, Map<N1, List<E>>> entries : newEdges.entrySet()) {
            Map<N1, List<E>> edges = entries.getValue();
            N1 v = entries.getKey();

            for (Map.Entry<N1, List<E>> edgesEntry : edges.entrySet()) {
                N1 w = edgesEntry.getKey();
                if (!allowSelfLoops && v.equals(w)) continue;

                E1 newEdge = verticesMap.apply(edgesEntry.getValue());
                result.insert(v, w, newEdge);
            }
        }

        return result;
    }
}

File: src/main/scala/org/gga/graph/transform/Morph.scala
package org.gga.graph.transform

import org.gga.graph.Graph
import org.gga.graph.MutableGraph
import org.gga.graph.impl.DataGraphImpl
import org.gga.graph.impl.SparseGraphImpl
import org.gga.graph.maps.DataGraph
import com.google.common.base.Preconditions.checkState
import collection.mutable

/**
 * @author mike
 */
object Morph {
  def morphGraph(g: Graph, map: (Int)=>Int): Graph = {
    val vertexMap: Array[Int] = new Array[Int](g.V)
    var newSize: Int = -1

    for (v <- g.vertices) {
      val t = map(v)
      vertexMap(v) = t
      newSize = newSize max t
    }

    newSize += 1
    checkState(newSize > 0)

    val edges: scala.collection.mutable.Set[(Int, Int)] = scala.collection.mutable.Set.empty
    val result: MutableGraph = new SparseGraphImpl(newSize, g.isDirected)

    for (v <- g.vertices) {
      for (e <- g.edges(v)) {
        val v1: Int = vertexMap(v)
        val w1: Int = vertexMap(e.w)

        val p = (v1, w1)
        if (!edges.contains(p)) {
          result.insert(v1, w1)
          edges.add(p)
        }
      }
    }

    result
  }

  def isomorph[N >: AnyRef, E](g: DataGraph[N, E], map: (Int) => Int): DataGraph[N, E] = {
    val vertexMap: Array[Int] = new Array[Int](g.V)
    var newSize: Int = -1

    for (v <- g.vertices) {
      val t: Int = map(v)
      vertexMap(v) = t
      newSize = newSize max t
    }

    newSize += 1
    checkState(newSize > 0 && newSize == g.V)

    val result: DataGraph[N, E] = new DataGraphImpl[N, E](newSize, g.isDirected)

    for (v <- g.vertices) {
      result.setNode(vertexMap(v), g.node(v))
    }

    for (v <- g.vertices) {
      for (e <- g.intGraph.edges(v)) {
        val v1: Int = vertexMap(v)
        val w1: Int = vertexMap(e.w)
        result.insert(v1, w1, g.edge(v, e.w).get)
      }
    }

    result
  }

  def morph[N, E, N1, E1](g: DataGraph[N, E], nodeMap: (N) => (N1), verticesMap: (E) => (E1)): DataGraph[N1, E1] = {
    val nodesDataMap: mutable.ArraySeq[N1] = new mutable.ArraySeq(g.V)

    for (v <- g.vertices) {
      nodesDataMap(v) = nodeMap(g.node(v))
    }

    val newEdges: mutable.Map[N1, mutable.Map[N1, mutable.Buffer[E]]] = mutable.Map.empty
    for (v <- g.vertices) {
      val n1: N1 = nodesDataMap(v)

      val edges = newEdges.getOrElseUpdate(n1, {mutable.Map.empty})

      for (e <- g.intGraph.edges(v)) {
        val n2: N1 = nodesDataMap(e.w)
        var edgesList = edges.getOrElseUpdate(n2, {mutable.Buffer.empty})
        edgesList += g.edge(e)
      }
    }


    // todo: finish
/*
    val newNodes: Array[N1] = newEdges.keySet.toArray
    val newSize: Int = newEdges.size
    val result: DataGraph[N1, E1] = new DataGraphImpl[N1, E1](newSize, g.isDirected)

    {
      var v: Int = 0
      while (v < newSize) {
        {
          result.setNode(v, newNodes(v).asInstanceOf[N1])
        }
        ({
          v += 1; v
        })
      }
    }
    for (n1 <- newEdges.keySet) {
      val edges: Map[N1, List[E]] = newEdges.get(n1)
      for (n2 <- edges.keySet) {
        val oldEdges: List[E] = edges.get(n2)
        val newEdge: E1 = verticesMap.fun(oldEdges)
        result.insert(n1, n2, newEdge)
      }
    }
    return result
*/
    throw new UnsupportedOperationException
  }
}